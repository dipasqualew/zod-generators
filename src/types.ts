/**
 * Represents a seed value
 */
export type Seed = number | boolean | null;

export type RNG = () => number;

/**
 * Recursively applies Partial
 * to all keys in T
 */
export type DeepPartial<T> = {
  [P in keyof T]?: DeepPartial<T[P]>;
};


/**
 * Returns an object that maps
 * all the keys of T to unknown
 */
export type RequiredKeys<T> = {
  [K in keyof Required<T>]: unknown;
};

/**
 * Returns an object that partially maps
 * all the keys of T to unknown
 */
export type PartialKeys<T> = {
  [K in keyof T]?: unknown
};


/**
 * Defines the shape of the options
 * that can be passed to a mock function
 */
export interface MockOptions<T> {
  override?: DeepPartial<T> | (() => DeepPartial<T>);
  seed?: Seed;
  rng?: RNG;
}

/**
 * Defines a collection of mocks
 */
export type Mocks<T> = Record<string, T>;

/**
 * Defines a mocks
 */
export type MocksGenerator = <T>() => Mocks<T>;
