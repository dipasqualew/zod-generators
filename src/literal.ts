import type { ZodLiteral } from 'zod';

import type { MockOptions } from './types';

/**
 * Generates valid literal mock
 * from the given ZodLiteral definition
 *
 * @param field
 * @param _options
 */
export const mockValid = <T>(field: ZodLiteral<T>, _options: MockOptions<T>) => {
  return {
    DEFAULT: field._def.value,
  };
};

/**
 * Generates invalid literal mocks
 * from the given ZodLiteral definition
 *
 * @param field
 * @param _options
 */
export const mockInvalid = <T>(field: ZodLiteral<T>, _options: MockOptions<T>) => {
  return {
    // This will catch all values
    // and transform it to something else
    DEFAULT: !field._def.value,
  };
};
