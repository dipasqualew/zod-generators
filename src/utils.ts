/**
 * Latin alphabet
 */
export const ALPHABET = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789'.split('');

export const MIN_INTEGER = Math.round(Number.MIN_SAFE_INTEGER / 3);
export const MAX_INTEGER = Math.round(Number.MAX_SAFE_INTEGER / 3);

/**
 * Returns a random element
 * from the given array
 *
 * @param origin
 */
export const getRandomElement = <T>(origin: T[]): T => {
  const index = Math.floor(Math.random() * origin.length);
  const item = origin[index];

  return item;
};

/**
 * Returns a random number between
 * the given min and the given max
 *
 * @param min
 * @param max
 * @param integer Whether the number should be an integer
 */
export const getRandomNumber = (min: number, max: number, integer = false): number => {
  const safeMin = Math.max(min, MIN_INTEGER);
  const safeMax = Math.min(max, MAX_INTEGER);

  const num = (Math.random() * (safeMax - safeMin)) + safeMin;

  if (integer) {
    return Math.floor(num);
  }

  return num;
};


/**
 * Returns a random string
 * of the given length
 *
 * @param length
 * @param alphabet
 */
export const getString = (length: number, alphabet = ALPHABET): string => {
  const elements = Array.from({ length }).map(() => getRandomElement(alphabet));

  return elements.join('');
};

/**
 * Returns a random string with
 * the given minLength and maxLength
 *
 * @param minLength
 * @param maxLength
 * @param alphabet
 */
export const getRandomString = (minLength: number, maxLength: number, alphabet = ALPHABET): string => {
  const length = getRandomNumber(minLength, maxLength, true);

  return getString(length, alphabet);
};
