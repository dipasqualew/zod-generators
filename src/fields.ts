/* eslint-disable no-use-before-define */
import * as zod from 'zod';

import * as literal from './literal';
import * as any from './primitives/any';
import * as bigint from './primitives/bigint';
import * as boolean from './primitives/boolean';
import * as date from './primitives/date';
import * as znull from './primitives/null';
import * as number from './primitives/number';
import * as string from './primitives/string';
import * as zundefined from './primitives/undefined';
import { DeepPartial, MockOptions } from './types';

/**
 * Returns the generator of valid values
 * for the given field
 *
 * @param field
 * @param options
 */
export const mockValid = <T extends zod.ZodAny>(field: T, options: MockOptions<zod.infer<T>> = {}): Record<string, zod.infer<T>> => {
  if (field instanceof zod.ZodObject) {
    return generateValidObjects(field as zod.ZodObject<zod.ZodRawShape>, options);
  }

  if ('override' in options && options.override !== undefined) {
    const value = options.override instanceof Function ? options.override() : options.override;

    return { DEFAULT: value };
  }

  if (field instanceof zod.ZodString) {
    return string.mockValid(field);
  }

  if (field instanceof zod.ZodNumber) {
    return number.mockValid(field);
  }

  if (field instanceof zod.ZodBoolean) {
    return boolean.mockValid(field);
  }

  if (field instanceof zod.ZodArray) {
    return generateValidArrays(field);
  }

  if (field instanceof zod.ZodDate) {
    return date.mockValid(field);
  }

  if (field instanceof zod.ZodBigInt) {
    return bigint.mockValid(field);
  }

  if (field instanceof zod.ZodLiteral) {
    return literal.mockValid(field);
  }

  if (field instanceof zod.ZodNull) {
    return znull.mockValid(field);
  }

  if (field instanceof zod.ZodUndefined || field instanceof zod.ZodVoid) {
    return zundefined.mockValid(field);
  }

  return any.mockValid(field);
};

/**
 * Returns the generator of invalid values
 * for the given field
 *
 * @param field
 * @param options
 */
export const mockInvalid = <T extends zod.ZodAny>(field: T, options: MockOptions<zod.infer<T>> = {}): Record<string, DeepPartial<zod.infer<T>>> => {
  if (field instanceof zod.ZodObject) {
    return generateInvalidObjects(field as zod.ZodObject<zod.ZodRawShape>, options);
  }

  if ('override' in options && options.override !== undefined) {
    const value = options.override instanceof Function ? options.override() : options.override;

    return { DEFAULT: value };
  }

  if (field instanceof zod.ZodString) {
    return string.mockInvalid(field);
  }

  if (field instanceof zod.ZodNumber) {
    return number.mockInvalid(field);
  }

  if (field instanceof zod.ZodBoolean) {
    return boolean.mockInvalid(field);
  }

  if (field instanceof zod.ZodArray) {
    return generateInvalidArrays(field);
  }

  if (field instanceof zod.ZodDate) {
    return date.mockInvalid(field);
  }

  if (field instanceof zod.ZodLiteral) {
    return literal.mockInvalid(field);
  }

  if (field instanceof zod.ZodBigInt) {
    return bigint.mockInvalid(field);
  }

  if (field instanceof zod.ZodNull) {
    return znull.mockInvalid(field);
  }

  if (field instanceof zod.ZodUndefined || field instanceof zod.ZodVoid) {
    return zundefined.mockInvalid(field);
  }

  return any.mockInvalid(field);
};

/**
 * Returns all mocks
 * for the given field
 *
 * @param field
 * @param options
 */
export const mock = <T extends zod.ZodAny>(field: T, options: MockOptions<zod.infer<T>> = {}) => {
  return {
    valid: mockValid(field, options),
    invalid: mockInvalid(field, options),
  };
};

/**
 * Generates valid arrays
 * from the given ZodArray definition
 *
 * @param field
 */
export const generateValidArrays = <T extends zod.ZodAny>(field: zod.ZodArray<T>) => {
  const items = mockValid(field._def.type);

  return {
    DEFAULT: Object.values(items),
  };
};

/**
 * Generates invalid arrays
 * from the given ZodArray definition
 *
 * @param field
 */
export const generateInvalidArrays = <T extends zod.ZodAny>(field: zod.ZodArray<T>): Record<string, DeepPartial<zod.TypeOf<T>>[]> => {
  const items = mockInvalid(field._def.type);

  const values = Object.values(items);

  if (values.length) {
    return {
      DEFAULT: values,
    };
  }

  return {};
};

/**
 * Generates valid objects
 * from the given ZodObject definition
 *
 * @param field
 * @param options
 */
export const generateValidObjects = <T extends zod.ZodRawShape>(field: zod.ZodObject<T>, options: MockOptions<unknown> = {}) => {
  type Shape = zod.infer<typeof field>;

  const defaultMock: Partial<Shape> = {};
  const variations: Record<string, Record<string, unknown>> = {};

  Object.entries(field.shape).forEach((entry) => {
    type Key = keyof Shape;
    type Value = Shape[keyof Shape];
    const key = entry[0] as Key;
    const def = entry[1] as Value;
    const newOptions = { ...options } as MockOptions<Value>;
    newOptions.override = ((options.override || {}) as Record<string, Value>)[entry[0]];

    const validMocks = mockValid(def, newOptions);

    defaultMock[key] = validMocks.DEFAULT as Shape[keyof Shape];
    variations[entry[0]] = validMocks;
  });

  const mocks = [
    ['DEFAULT', defaultMock],
  ];

  Object.entries(variations).forEach((variation) => {
    Object.entries(variation[1]).forEach((entry) => {
      const name = `${variation[0]}__${entry[0]}`;
      const value = { ...defaultMock, [variation[0]]: entry[1] };

      mocks.push([name, value]);
    });
  });

  return Object.fromEntries(mocks) as { DEFAULT: Shape } & Record<string, Shape>;
};

/**
 * Generates valid objects
 * from the given ZodObject definition
 *
 * @param field
 * @param options
 */
export const generateInvalidObjects = <T extends zod.ZodRawShape>(field: zod.ZodObject<T>, options: MockOptions<unknown> = {}) => {
  type Shape = zod.infer<typeof field>;

  const valid = generateValidObjects(field).DEFAULT;

  const defaultMock: Partial<Shape> = {};
  const variations: Record<string, Record<string, unknown>> = {};

  Object.entries(field.shape).forEach((entry) => {
    type Key = keyof Shape;
    type Value = Shape[keyof Shape];
    const key = entry[0] as Key;
    const def = entry[1] as Value;
    const newOptions = { ...options } as MockOptions<Value>;
    newOptions.override = ((options.override || {}) as Record<string, Value>)[entry[0]];

    const invalidMocks = mockInvalid(def, newOptions);

    if ('DEFAULT' in invalidMocks) {
      defaultMock[key] = invalidMocks.DEFAULT as Shape[keyof Shape];
      variations[entry[0]] = invalidMocks;
    }
  });

  const mocks: [string, DeepPartial<Shape>][] = [];

  if (Object.keys(defaultMock).length) {
    mocks.push(['DEFAULT', defaultMock]);
  }

  Object.entries(variations).forEach(([fieldName, fieldVariations]) => {
    Object.entries(fieldVariations).forEach(([description, fieldValue]) => {
      const name = `${fieldName}__${description}`;
      const value = { ...valid, [fieldName]: fieldValue };

      mocks.push([name, value]);
    });
  });

  return Object.fromEntries(mocks) as { DEFAULT: DeepPartial<Shape> } & Record<string, DeepPartial<Shape>>;
};
