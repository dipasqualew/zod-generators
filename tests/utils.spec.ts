import { getUUID, mulberry32 } from '../src/utils';

describe('utils', () => {
  describe('mulberry32', () => {
    it('returns random numbers', () => {
      const rng = mulberry32(1);
      const length = 10000;
      const arr = Array.from({ length }).map(() => rng());
      const set = new Set(arr);
      expect(set.size).toEqual(length);
    });

    it('instances from the same seed return the same sequence', () => {
      const rng1 = mulberry32(1);
      const rng2 = mulberry32(1);

      const length = 500;

      const set = new Set();

      for (let i = 0; i < length; i += 1) {
        set.add(rng1());
        set.add(rng2());

        expect(set.size).toEqual(i + 1);
      }
    });

    it('instances from different seeds return the different sequences', () => {
      const rng1 = mulberry32(1);
      const rng2 = mulberry32(2);

      const length = 500;

      const set = new Set();

      for (let i = 0; i < length; i += 1) {
        set.add(rng1());
        set.add(rng2());

        expect(set.size).toEqual((i + 1) * 2);
      }
    });
  });

  describe('getUUID', () => {
    describe('without a seed', () => {
      it('returns a random UUID', () => {
        const length = 500;
        const arr = Array.from({ length }).map(() => getUUID());
        const set = new Set(arr);

        expect(set.size).toEqual(length);
      });
    });

    describe('with a seed', () => {
      it('returns a sequence of deterministic UUIDs', () => {
        const rng = mulberry32(1);
        const uuids = [
          getUUID(rng),
          getUUID(rng),
          getUUID(rng),
        ];

        expect(uuids).toMatchSnapshot();
      });
    });
  });
});
