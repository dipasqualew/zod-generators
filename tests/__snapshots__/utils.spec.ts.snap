// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils getUUID with a seed returns a sequence of deterministic UUIDs 1`] = `
Array [
  "c994787c-3b4c-5e2f-81de-33ab9b048e46",
  "2bd4f558-ee91-5183-bcf8-c5663aa32b54",
  "9d6fc420-dfc7-5aeb-838c-4b94e6d797f2",
]
`;
